<a href="https://colab.research.google.com/github/EdioHub/Score_Credito/blob/main/Score_cr%C3%A9dito.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Projeto Python IA: Inteligência Artificial e Previsões

### Case: Score de Crédito de Clientes  de um banco

 Análise de todos os clientes do banco para a criação de modelos que consigam ler as informações do cliente e dizer automaticamente o score de crédito dele: Ruim, Ok, Bom

 Uso de modelos comparativos entre KNN, Naives, SVC, Ada, Forest, Tree


In [5]:
import pandas as pd


In [6]:
import numpy as np
import sklearn

In [ ]:
caminho = ('/content/drive/MyDrive/Ciência de dados Python/Python insights-Machine fazendo/Aula 3/drive-download-20230728T232828Z-001/clientes.csv')

In [7]:
data = pd.read_csv('/content/cancelamentos.csv')
data.sample(5)

CustomerID  idade    sexo  tempo_como_cliente  frequencia_uso  \
156569    161380.0   63.0    Male                29.0            12.0   
205839    210917.0   63.0    Male                49.0            28.0   
193554    198365.0   65.0  Female                 8.0             7.0   
204417    209441.0   57.0  Female                 3.0            11.0   
2705        2714.0   51.0  Female                 3.0             6.0   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
156569                  4.0         18.0      Basic        Quarterly   
205839                  3.0          9.0   Standard          Monthly   
193554                  7.0         17.0      Basic          Monthly   
204417                  2.0         20.0    Premium           Annual   
2705                    4.0         15.0   Standard        Quarterly   

        total_gasto  meses_ultima_interacao  cancelou  
156569       411.00                    13.0       1.0  
205839       479.52                     7.0       1.0  
193554       750.00                    17.0       1.0  
204417       415.06                    21.0       1.0  
2705         620.00                    18.0       1.0

In [ ]:
data['profissao'].unique()

array(['cientista', 'professor', 'engenheiro', 'empresario',
       'desenvolvedor', 'advogado', 'gerente_midia', 'medico',
       'jornalista', 'gerente', 'contador', 'musico', 'mecanico',
       'escritor', 'arquiteto'], dtype=object)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 25 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   id_cliente                100000 non-null  int64  
 1   mes                       100000 non-null  int64  
 2   idade                     100000 non-null  float64
 3   profissao                 100000 non-null  object 
 4   salario_anual             100000 non-null  float64
 5   num_contas                100000 non-null  float64
 6   num_cartoes               100000 non-null  float64
 7   juros_emprestimo          100000 non-null  float64
 8   num_emprestimos           100000 non-null  float64
 9   dias_atraso               100000 non-null  float64
 10  num_pagamentos_atrasados  100000 non-null  float64
 11  num_verificacoes_credito  100000 non-null  float64
 12  mix_credito               100000 non-null  object 
 13  divida_total              100000 non-null  fl

In [ ]:
data.describe()

id_cliente            mes          idade  salario_anual  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean    25982.666640       4.500000      33.316340   50505.123449   
std     14340.543051       2.291299      10.764812   38299.422093   
min      1006.000000       1.000000      14.000000    7005.930000   
25%     13664.500000       2.750000      24.000000   19342.972500   
50%     25777.000000       4.500000      33.000000   36999.705000   
75%     38385.000000       6.250000      42.000000   71683.470000   
max     50999.000000       8.000000      56.000000  179987.280000   

          num_contas    num_cartoes  juros_emprestimo  num_emprestimos  \
count  100000.000000  100000.000000      100000.00000    100000.000000   
mean        5.368820       5.533570          14.53208         3.532880   
std         2.593314       2.067098           8.74133         2.446356   
min         0.000000       0.000000           1.00000         0.000000   
25%         3.000000       4.000000           7.00000         2.000000   
50%         5.000000       5.000000          13.00000         3.000000   
75%         7.000000       7.000000          20.00000         5.000000   
max        11.000000      11.000000          34.00000         9.000000   

        dias_atraso  num_pagamentos_atrasados  ...   divida_total  \
count  100000.00000             100000.000000  ...  100000.000000   
mean       21.08141                 13.313120  ...    1426.220376   
std        14.80456                  6.237166  ...    1155.129026   
min         0.00000                  0.000000  ...       0.230000   
25%        10.00000                  9.000000  ...     566.072500   
50%        18.00000                 14.000000  ...    1166.155000   
75%        28.00000                 18.000000  ...    1945.962500   
max        62.00000                 25.000000  ...    4998.070000   

       taxa_uso_credito  idade_historico_credito  investimento_mensal  \
count     100000.000000            100000.000000        100000.000000   
mean          32.285173               221.220460            55.101315   
std            5.116875                99.680716            39.006932   
min           20.000000                 1.000000             0.000000   
25%           28.052567               144.000000            27.959111   
50%           32.305784               219.000000            45.156550   
75%           36.496663               302.000000            71.295797   
max           50.000000               404.000000           434.191089   

       saldo_final_mes  emprestimo_carro  emprestimo_casa  emprestimo_pessoal  \
count    100000.000000     100000.000000    100000.000000       100000.000000   
mean        392.697586          0.305600         0.511760            0.311040   
std         201.652719          0.460663         0.499864            0.462921   
min           0.007760          0.000000         0.000000            0.000000   
25%         267.615983          0.000000         0.000000            0.000000   
50%         333.865366          0.000000         1.000000            0.000000   
75%         463.215683          1.000000         1.000000            1.000000   
max        1183.930696          1.000000         1.000000            1.000000   

       emprestimo_credito  emprestimo_estudantil  
count        100000.00000           100000.00000  
mean              0.31728                0.31040  
std               0.46542                0.46266  
min               0.00000                0.00000  
25%               0.00000                0.00000  
50%               0.00000                0.00000  
75%               1.00000                1.00000  
max               1.00000                1.00000  

[8 rows x 21 columns]

In [ ]:
from sklearn.preprocessing import LabelEncoder
codificador = LabelEncoder()

In [ ]:
data['profissao']=codificador.fit_transform(data['profissao'])
data['mix_credito']=codificador.fit_transform(data['mix_credito'])
data['comportamento_pagamento']=codificador.fit_transform(data['comportamento_pagamento'])


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 25 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   id_cliente                100000 non-null  int64  
 1   mes                       100000 non-null  int64  
 2   idade                     100000 non-null  float64
 3   profissao                 100000 non-null  int64  
 4   salario_anual             100000 non-null  float64
 5   num_contas                100000 non-null  float64
 6   num_cartoes               100000 non-null  float64
 7   juros_emprestimo          100000 non-null  float64
 8   num_emprestimos           100000 non-null  float64
 9   dias_atraso               100000 non-null  float64
 10  num_pagamentos_atrasados  100000 non-null  float64
 11  num_verificacoes_credito  100000 non-null  float64
 12  mix_credito               100000 non-null  int64  
 13  divida_total              100000 non-null  fl

In [ ]:
y = data['score_credito']
x = data.drop(['score_credito','id_cliente'], axis = 1)

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3)

In [ ]:
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier

In [ ]:
model_forest = RandomForestClassifier()
model_knn = KNeighborsClassifier()
model_naives = GaussianNB()
model_tree = DecisionTreeClassifier()
model_svc = SVC()
model_ada = AdaBoostClassifier()
model_gausian = GaussianProcessClassifier()

In [ ]:
model_forest.fit(x_train, y_train)


RandomForestClassifier()

In [ ]:
model_knn.fit(x_train, y_train)

KNeighborsClassifier()

In [ ]:
model_naives.fit(x_train, y_train)

GaussianNB()

In [ ]:
model_tree.fit(x_train, y_train)

DecisionTreeClassifier()

In [ ]:
model_ada.fit(x_train, y_train)

AdaBoostClassifier()

In [ ]:
model_svc.fit(x_train, y_train)

SVC()

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
previsao_forest = model_forest.predict(x_test)
previsao_knn = model_knn.predict(x_test)
previsao_naives = model_naives.predict(x_test)
previsao_tree = model_tree.predict(x_test)
previsao_ada = model_ada.predict(x_test)
previsao_svc = model_svc.predict(x_test)

In [ ]:
print('forest = ', accuracy_score(y_test,previsao_forest))
print('Knn = ', accuracy_score(y_test,previsao_knn))
print('naives = ', accuracy_score(y_test,previsao_naives))
print('tree = ', accuracy_score(y_test,previsao_tree))
print('ada = ', accuracy_score(y_test,previsao_ada))
print('svc = ', accuracy_score(y_test,previsao_svc))

forest =  0.8221666666666667
Knn =  0.7341
naives =  0.5981666666666666
tree =  0.7313333333333333
ada =  0.6445333333333333
svc =  0.5354666666666666


# Conclusão

Com base nos resultados obtidos a partir da avaliação de diversos modelos de aprendizado de máquina, fica evidente que o Modelo Forest se destacou como a escolha mais eficaz para este sistema preditivo. Com uma taxa de acurácia de aproximadamente 82.22%, o Modelo Forest demonstrou sua capacidade de realizar previsões com alta precisão, tornando-se uma ferramenta valiosa para tomar decisões embasadas.

Comparativamente, outros modelos como K-Nearest Neighbors (Knn), Decision Tree, AdaBoost e Naive Bayes apresentaram taxas de acurácia inferiores, variando entre 73.41% e 59.82%. Isso ressalta a superioridade do Modelo Forest no contexto deste estudo específico.

No entanto, é importante ressaltar que o Support Vector Classifier (SVC) não obteve um desempenho satisfatório, atingindo uma taxa de acurácia de apenas 53.55%. Além disso, observou-se que o SVC demonstrou uma notável lentidão devido à complexidade de sua modelagem. Esse fator, aliado ao baixo desempenho, torna o SVC inadequado para ser utilizado em sistemas que lidam com bancos de dados de grande porte.

Diante desses resultados, conclui-se que a seleção do modelo de aprendizado de máquina é crucial para alcançar previsões precisas e eficientes. O Modelo Forest se sobressaiu, oferecendo um equilíbrio entre acurácia e desempenho, sendo altamente recomendado para a implementação neste sistema preditivo. Por outro lado, o uso do SVC deve ser evitado, especialmente em cenários que demandam rapidez e lidam com volumes significativos de dados. Este estudo reforça a importância de considerar não apenas a acurácia, mas também fatores como eficiência e escalabilidade ao escolher um modelo de aprendizado de máquina para tarefas preditivas.